kroxylicious-sample/src/test/java/io/kroxylicious/sample/SampleFilterIntegrationTest.java

```diff
@@ -0,0 +1,154 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+@ExtendWith(KafkaClusterExtension.class)
+public class SampleFilterIntegrationTest {
+
+    private static final String PRE_TRANSFORM_VALUE = "foo";
+    private static final String NO_TRANSFORM_VALUE = "sample";
+    private static final String PRODUCE_TRANSFORM_VALUE = "bar";
+    private static final String FETCH_TRANSFORM_VALUE = "baz";
+    private static final String FIND_CONFIG_FIELD = "findValue";
+    private static final String REPLACE_CONFIG_FIELD = "replaceValue";
+    private static final Map<String, Object> PRODUCE_CONFIG = Map.of(FIND_CONFIG_FIELD, "foo", REPLACE_CONFIG_FIELD, "bar");
+    private static final Map<String, Object> FETCH_CONFIG = Map.of(FIND_CONFIG_FIELD, "bar", REPLACE_CONFIG_FIELD, "baz");
+    private static final Integer TIMEOUT_SECONDS = 10;
+    private static final String TOPIC_NAME = "test";
+    private static final Integer TOPIC_PARTITIONS = 1;
+    private static final Short TOPIC_REPLICATION = 1;
+
+    KafkaCluster cluster;
+    private KroxyliciousTester tester;
+
+    @BeforeEach
+    public void beforeEach() {
+        tester = kroxyliciousTester(withDefaultFilters(proxy(cluster))
+                .addToFilters(new FilterDefinitionBuilder(SampleContributor.SAMPLE_PRODUCE).withConfig(PRODUCE_CONFIG).build())
+                .addToFilters(new FilterDefinitionBuilder(SampleContributor.SAMPLE_FETCH).withConfig(FETCH_CONFIG).build()));
+    }
+
+    @AfterEach
+    public void afterEach() {
+        tester.close();
+    }
+
+    @Test
+    public void sampleFilterWillTransformRoundTripTest(Admin admin) {
+        try (Producer<String, String> producer = tester.producer();
+                Consumer<String, byte[]> kafkaClusterConsumer = new KafkaConsumer<>(mergeMaps(cluster.getKafkaClientConfiguration(),
+                        Map.of(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class, ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+                                ByteArrayDeserializer.class, ConsumerConfig.GROUP_ID_CONFIG, "group-id-0", ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")));
+                Consumer<String, byte[]> proxyConsumer = tester.consumer(Serdes.String(), Serdes.ByteArray(),
+                        Map.of(ConsumerConfig.GROUP_ID_CONFIG, "group-id-1", ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest"))) {
+            admin.createTopics(List.of(new NewTopic(TOPIC_NAME, TOPIC_PARTITIONS, TOPIC_REPLICATION))).all().get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            producer.send(new ProducerRecord<>(TOPIC_NAME, PRE_TRANSFORM_VALUE)).get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            // get from cluster
+            kafkaClusterConsumer.subscribe(List.of(TOPIC_NAME));
+            ConsumerRecords<String, byte[]> kafkaClusterPoll = kafkaClusterConsumer.poll(Duration.ofSeconds(TIMEOUT_SECONDS));
```

Comment:
I see why you're doing this, and I like it in principal. However I hesitate as the sample filter is meant to be an exemplar and I'm not convinced we want to establish polling the underlying cluster directly as the default expectation. 

I'm wondering if it would be better to configure the proxy with only a `ProduceFilter` and verify via the tester that the consumed value is transformed or not (as appropriate). Then in an additional test verify the reverse with only a `FetchResponseFilter`?

Thoughts @franvila @k-wall @robobario ?

Reply:
Yeahp, in this case we should be able to observe both behaviours via Kroxylicious and not worry too much about how it's stored in the backing cluster. If I produce `xyzfoobar` and consume `xyzbazbaz`, we should know that both filters did their job.

But I think that often I would want to check what is in the real cluster. Like for topic encryption, the observed behaviour of kroxylicious as a black box is the same as if there was no encryption (it gives back the same bytes you put in). So we need to check what is sent to the real cluster. I think this is going to be a common thing and worth having an example of.

Reply:
>But I think that often I would want to check what is in the real cluster. Like for topic encryption, the observed behaviour of kroxylicious as a black box is the same as if there was no encryption (it gives back the same bytes you put in). So we need to check what is sent to the real cluster. I think this is going to be a common thing and worth having an example of.

Yeah I kinda agree but don't we get the same by including asymmetric filters rather than hitting the underlying cluster directly? (I know I have argued for underlying cluster tests in the past and I think those scenario's could probably have been handled with asymmetric filter deployment)

Reply:
> I'm wondering if it would be better to configure the proxy with only a `ProduceFilter` and verify via the tester that the consumed value is transformed or not (as appropriate). Then in an additional test verify the reverse with only a `FetchResponseFilter`?

I agree splitting the tests as much as possible and not trying to check many things in the same test. With that we will be sure that if we found a bug, it is clear where it is, without any further investigation.

Reply:
I agree with having separate tests per filter, and the latest commit splits them up as well as moving a bunch of the common code into it's own class for easier reuse (i.e. if we add more sample filters later).